business/istio_status.go

```diff
@@ -271,26 +271,35 @@ func (iss *IstioStatusService) getIstiodReachingCheck() (IstioComponentStatus, e
 		}
 	}
 
+	// Fetch pod's IPs
+	istiodIPs := make([]string, 0, len(healthyIstiods))
+	for _, istiod := range healthyIstiods {
+		if istiod.Status.PodIP != "" {
+			istiodIPs = append(istiodIPs, istiod.Status.PodIP)
+		}
+	}
+
 	wg := sync.WaitGroup{}
 	wg.Add(len(healthyIstiods))
 	syncChan := make(chan ComponentStatus, len(healthyIstiods))
 
-	for _, istiod := range healthyIstiods {
-		go func(name, namespace string) {
+	for i, istiod := range healthyIstiods {
+		go func(name, namespace, ip string) {
 			defer wg.Done()
 			// Using the proxy method to make sure that K8s API has access to the Istio Control Plane namespace.
 			// By proxying one Istiod, we ensure that the following connection is allowed:
 			// Kiali -> K8s API (proxy) -> istiod
 			// This scenario is no obvious for private clusters (like GKE private cluster)
-			_, err := iss.k8s.GetPodProxy(namespace, name, "/ready")
-			if err != nil {
+			debugURI := fmt.Sprintf("http://%s:8080/ready", ip)
```

Comment:
We should discuss this.
Port 8080 is deprecated according to [1]

[1] https://istio.io/latest/docs/ops/deployment/requirements/#ports-used-by-istio

Reply:
Just looking at the endpoint:

```
$ kubectl get endpoints/istiod -n istio-system -o yaml
apiVersion: v1
kind: Endpoints
metadata:
  annotations:
    endpoints.kubernetes.io/last-change-trigger-time: "2021-06-22T07:27:14Z"
  creationTimestamp: "2021-06-22T07:27:07Z"
  labels:
    app: istiod
    install.operator.istio.io/owning-resource: unknown
    install.operator.istio.io/owning-resource-namespace: istio-system
    istio: pilot
    istio.io/rev: default
    operator.istio.io/component: Pilot
    operator.istio.io/managed: Reconcile
    operator.istio.io/version: 1.10.1
    release: istio
  name: istiod
  namespace: istio-system
  resourceVersion: "608"
  uid: ce94b645-2fe2-4ffa-b49a-6280fb5000d7
subsets:
- addresses:
  - ip: 172.17.0.5
    nodeName: minikube
    targetRef:
      kind: Pod
      name: istiod-85c958cd6-kcfk9
      namespace: istio-system
      resourceVersion: "606"
      uid: f52827e2-2726-4846-ac22-943720e83452
  ports:
  - name: https-dns
    port: 15012
    protocol: TCP
  - name: grpc-xds
    port: 15010
    protocol: TCP
  - name: https-webhook
    port: 15017
    protocol: TCP
  - name: http-monitoring
    port: 15014
    protocol: TCP

```

Reply:
I know. However, I haven't found the way to re-use the same readiness probes used for kubernetes to manage pod's life-cycle. It looks like is deprecated for the debug endpoints (/debug/[syncz|registryz|...]), but they still use it for readiness probes. I couldn't find another way to fetch that endpoint otherwise.

We could use another endpoint to consume from the 15014 port. However, I don't know if it would be as light and accurate as this one.

```
$ kubectl get pod -n istio-system istiod-12312882 -o yaml
apiVersion: v1
kind: Pod
metadata:
  name: istiod-8546f98dcb-qwhzx
  namespace: istio-system
  annotations:
    prometheus.io/port: "15014"
    prometheus.io/scrape: "true"
    sidecar.istio.io/inject: "false"
...
    readinessProbe:
      failureThreshold: 3
      httpGet:
        path: /ready
        port: 8080
        scheme: HTTP
      initialDelaySeconds: 1
      periodSeconds: 3
      successThreshold: 1
      timeoutSeconds: 5
....
```